File: 2906-construct-product-matrix/2906-construct-product-matrix.cpp
class Solution {
public:
    vector<vector<int>> constructProductMatrix(vector<vector<int>>& grid) {
        
        vector<int> temp ;

        for( int i = 0 ; i < grid.size() ; i++ ){
            for( int j = 0 ; j < grid[0].size() ; j++ ){
                temp.push_back(grid[i][j]%12345) ;
            }
        }

        vector<int> prefix( temp.size() , 1 );
        vector<int> suffix( temp.size() , 1 );

        int multi = 1 ;
        for( int i = 0 ; i < temp.size() ; i++ ){
            multi = (multi * temp[i]) % 12345;
            prefix[i] = multi ;
        }

        multi = 1 ;
        for( int i = temp.size() - 1 ; i > -1 ; i-- ){
            multi = (multi * temp[i]) % 12345;
            suffix[i] = multi ;
        }

        vector<vector<int>> result( grid.size() , vector<int> ( grid[0].size() , 1 ));
        int k = 0 ;

        for( int i = 0 ; i < grid.size() ; i++ ){
            for( int j = 0 ; j < grid[0].size() ; j++ ){
                if( k == 0 ){
                    result[i][j] = suffix[k+1] % 12345 ;
                }
                else if( k == temp.size() - 1 ){
                    result[i][j] = prefix[k-1] % 12345 ;
                }
                else{
                    result[i][j] = ( prefix[k-1] * suffix[k+1] ) % 12345 ;
                }
                k++;
            }
        }

        return result ;
    }
};

File: 2906-construct-product-matrix/2906-construct-product-matrix.js
/**


 * @param {number[][]} grid
 * @return {number[][]}
 */
var constructProductMatrix = function(grid) {
    
    let MOD = 12345 
    let temp = []

    for( let i = 0 ; i < grid.length ; i++ ){
        for( let j = 0 ; j < grid[0].length ; j++ ){
            temp.push( grid[i][j] % MOD )
        }
    }

    let prefix = []
    let suffix = []
    let multi = 1 ;

    for( let i = 0 ; i < temp.length ; i++ ){
        multi = ( multi * temp[i] ) % MOD ;
        prefix.push(multi)
    }

    multi = 1 
    for( let i = temp.length - 1 ; i > -1 ; i-- ){
         multi = ( multi * temp[i] ) % MOD ;
         suffix[i] = multi ;
    }

    let result = []
    let k = 0 

    for( let i = 0 ; i < grid.length ; i++ ){

        let arr = []
        for( let j = 0 ; j < grid[0].length ; j++ ){
            if( k == 0 ){
                arr[j] =  suffix[k+1] % MOD 
            }
            else if( k == temp.length - 1 ){
                arr[j] =  prefix[k-1] % MOD 
            }
            else{
                arr[j] = ( prefix[k-1] * suffix[k+1] ) % MOD 
            }
            k++;
        }

        result.push(arr)
    }

    return result 

};